kv: test eviction on first range gossip

tbg opened the pull request:
Fixes #8164.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8288)

<!-- Reviewable:end -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender_test.go, line 626 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFYNnitr3Mp11KMzBl:-KOFYNnitr3Mp11KMzBm:1068954415) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L626)):*
> ```Go
> 	) {
> 		if len(key) == 0 {
> 			fmt.Println("lookup")
> ```

Delete?

---

*[kv/dist_sender_test.go, line 644 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFYzenKPQtLZMChSm_:-KOFYzenKPQtLZMChSma:-14366275) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L644)):*
> ```Go
> 	if _, pErr := client.SendWrapped(ds, context.Background(), getArg); pErr != nil {
> 		t.Fatal(pErr)
> 	}
> ```

Does the test need this part? I would think the `LookupRangeDescriptor` will force the caching. I'm probably just not understanding what this is doing.

---

*[kv/dist_sender_test.go, line 656 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFYFb7AocCbnf5Y7xX:-KOFYFb7AocCbnf5Y7xY:241065947) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L656)):*
> ```Go
> 	call()
> 	call()
> 	call()
> ```

Add a comment: perform multiple calls and check that the first-range is only looked up once. The subsequent calls will hit the cache. It would be mildly better to check `numFirstRange` after the first call.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-:-KOFZ6tcp1S0MGQRgurd:1417088721)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender_test.go, line 597 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFe_0vfrUAFhCa_fO-:-KOFe_0wWUl7VB7Gyx6D:-1555492316) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L597)):*
> ```Go
> 	defer leaktest.AfterTest(t)()
> 
> 	n := simulation.NewNetwork(1, true)
> ```

oh let's not do this plz

see #8285

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod comments

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[kv/dist_sender_test.go, line 602 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFn8JsyFXv0ewF9PVa:-KOFn8Jtr49ya8vIzTLw:-442962004) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L602)):*
> ```Go
> 	g := n.Nodes[0].Gossip
> 
> 	var sender client.SenderFunc = func(_ context.Context, ba roachpb.BatchRequest) (
> ```

`sender := `?

---

*[kv/dist_sender_test.go, line 621 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFnD2lxdsv9ZJvsagZ:-KOFnD2lxdsv9ZJvsag_:-1629528554) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L621)):*
> ```Go
> 
> 	var numFirstRange int32
> ```

rm this line

---

*[kv/dist_sender_test.go, line 675 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFnfuZU5x0b8ZrgjvB:-KOFnfuZU5x0b8ZrgjvC:5174784) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L675)):*
> ```Go
> 		return nil
> 	})
> ```

rm

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-:-KOFniMAH7Y70I9-w4fV:-477825471)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Someone wanna take this over? Otherwise will sit for 2wks.

On Wed, Aug 3, 2016, 09:30 Tamir Duberstein <notifications@github.com>
wrote:

> [image: :lgtm:]
> <https://camo.githubusercontent.com/41b3ec3419116e3b3f84507ad965646ce4ce1f0c/68747470733a2f2f72657669657761626c652e696f2f6c67746d2e706e67>
> mod comments
> ------------------------------
>
> Reviewed 1 of 1 files at r1.
> Review status: all files reviewed at latest revision, 7 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *kv/dist_sender_test.go, line 602 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFn8JsyFXv0ewF9PVa:-KOFn8Jtr49ya8vIzTLw:-442962004>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L602>):*
>
>   g := n.Nodes[0].Gossip
>
>   var sender client.SenderFunc = func(_ context.Context, ba roachpb.BatchRequest) (
>
> sender :=?
> ------------------------------
>
> *kv/dist_sender_test.go, line 621 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFnD2lxdsv9ZJvsagZ:-KOFnD2lxdsv9ZJvsag_:-1629528554>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L621>):*
>
>
>   var numFirstRange int32
>
> rm this line
> ------------------------------
>
> *kv/dist_sender_test.go, line 675 [r1]
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFnfuZU5x0b8ZrgjvB:-KOFnfuZU5x0b8ZrgjvC:5174784>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L675>):*
>
>       return nil
>   })
>
> rm
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-:-KOFniMAH7Y70I9-w4fV:-477825471>*
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8288#issuecomment-237252998>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AE135P-btuAvMK4V58vpFnb-jKByrdDgks5qcLDdgaJpZM4Jbn6w>
> .
>
-- 

-- Tobias

petermattis commented:
@tamird Can you pick this up as well?

tamird commented:
Sure.

On Wed, Aug 3, 2016 at 9:53 PM, Peter Mattis <notifications@github.com>
wrote:

> @tamird <https://github.com/tamird> Can you pick this up as well?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8288#issuecomment-237430224>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPMdL5GDy6_8V2uAeWWUmtUwTuRUVks5qcUYGgaJpZM4Jbn6w>
> .
>

tschottdorf commented:
@tamird is this obsolete by now or still yours? If the latter, might want to close this and open it under your own name when you get around to it.

tamird commented:
Oops, slipped through the cracks - all yours.

tbg committed:
kv: test eviction on first range gossip

I briefly stress tested this to make sure it isn't obviously flaky, and also
verified that it fails without #8163.

Fixes #8164.

tschottdorf commented:
PTAL @tamird

---

Review status: 0 of 7 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[kv/dist_sender_test.go, line 597 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFe_0vfrUAFhCa_fO-:-KPsK-LHHJEgADwQwAFq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L597)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> oh let's not do this plz
> 
> see #8285
</details>
Done.

---

*[kv/dist_sender_test.go, line 602 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFn8JsyFXv0ewF9PVa:-KPsKGFxh_ZrvPUNn_YM:bbwd37u) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L602)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `sender := `?
</details>
Done, though that incurs a `client.SenderFunc` cast below.

---

*[kv/dist_sender_test.go, line 621 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFnD2lxdsv9ZJvsagZ:-KPsKOjw8FPJlratnM_t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L621)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> rm this line
</details>
Done.

---

*[kv/dist_sender_test.go, line 626 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFYNnitr3Mp11KMzBl:-KPsKPG704oYE3jNaLmo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L626)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Delete?
</details>
Done.

---

*[kv/dist_sender_test.go, line 644 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFYzenKPQtLZMChSm_:-KPsM2s1dKv0ISvIWqyy:b-unyvxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L644)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Does the test need this part? I would think the `LookupRangeDescriptor` will force the caching. I'm probably just not understanding what this is doing.
</details>
No, you're right. Removed.

---

*[kv/dist_sender_test.go, line 656 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFYFb7AocCbnf5Y7xX:-KPsKjO8KwnNjiQgceP7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L656)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Add a comment: perform multiple calls and check that the first-range is only looked up once. The subsequent calls will hit the cache. It would be mildly better to check `numFirstRange` after the first call.
</details>
Done.

---

*[kv/dist_sender_test.go, line 675 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-KOFnfuZU5x0b8ZrgjvB:-KPsKlnn1ER4V9oJ4VOo:brneu7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/44b02f44f62b402b4d786d04005e491c1f00ab49/kv/dist_sender_test.go#L675)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> rm
</details>
Done. I really do these a lot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-:-KPsM5YjVu7U6fwOcxJ-:b-kb5iep)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 7 of 7 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8288#-:-KOIFGRRm-03vXMi1Emj:bnfp4nl)*
<!-- Sent from Reviewable.io -->
